regex engine - rework node sizing logic and related functionality

demerphq opened the pull request:
Currently there is a lot of brittle and opaque logic related to regex regnode (aka regop) sizing. 

* many objects share common node structures structures (eg, multiple different regnodes use `struct regnode_1`) and related macros.
*  there is a very confusing concept called `NEXTOPER(p)` heavily used in the regcomp.c (and somewhat used in regexec.c), this function looks like it does something like "find the next node in the compiled program which follows p", but in fact resolves to `p+1`, and is often used like this: `NEXTOPER(NEXTOPER(p))` to compute `p+2` for cases where the code assumes the regnode is of size 2. 
*  there is another concept called `regnext` which is a different kind of "next operator" for a node. The similarity in names with `NEXTOPER()` make it quite confusing what the difference is.
*  We have a utility array called `PL_regarglen` which can be used to look up how big a given opcode should be, but it does not handle all node types, only those of a constant size.
* There are other macros which can be used to find how big a variable length regnode is, but there is nothing programmatic that knows how to handle all regnode types.
* There is no programmatic way to know if a given regnode is variable length, it is only implied by the code. 

All of these factors make it difficult and error prone to adjust the sizes of regnodes in the compiled regex program and add new functionality, and understand the code.

This patch sequence attempts to resolve these issues: 

* Adds a new function `Perl_regnode_after()` which can be used to reliably find the regnode after a given regnode regardless of its type. 
* Replaces `NEXTOPER()` with `REGNODE_AFTER()` (and a set of specialty variants of `REGNODE_AFTER()`) which should distinguish it from `regnext()`.
* Adds logic to validate the special case variants of `REGNODE_AFTER()` which are used for performance reasons (no need to do run time array looks when we know the result will be zero at compile time).
* Adds a new per-nodetype typedef so that code that it is easy to distinguish between nodes types in the code. For instance if the code is operating on a structure used by the `ANYOF` operator then it can use the typedef `tregnode_ANYOF` instead of `struct charclass` which is shared with multiple other regnodes so that when the given implementation of a node type changes it is easy to find the relevant logic.
* Rename all the `PL_reg` vars which track meta-data about the regnode structures so that they start `PL_regnode_` and add underbars to make them more readable.
* Adds  a new regnode metadata variable `PL_regnode_arg_len_varies[]` which can be used to identify node types whose length is reflected by their internal state. 
* Moves `Perl_regnext()` into reginline.h along with `Perl_regnode_after()` and `Perl_check_regnode()`. The idea being that reginline.h contains inline functions which are strictly used by the regex engine only.

Various sundry changes to facilitate the above changes. 

With these patches it is relatively easy to change the definition of a given regnode and to be able to find the code related to it, and detect what code has broken because it makes unwarranted assumptions about regnode sizes or types. Before this PR when sizes went wrong a bunch of pointer math would end up producing the wrong answer and mayhem would ensue, but nary an assert would fire to show where things went wrong. With it the issues are obvious.  -- This PR *doesnt* include any regnode sizing changes, but a follow up patch does, and these comments are based on experience of resizing the BRANCH nodetype, which is one of the weirder edge cases, and this patch was essential to allowing that to be possible.

The changes in this PR inter mesh with each other. For instance, adding typedefs means we can provide REGNODE_AFTER_type(p,type) which essentially turns into p+=sizeof(type) for the cases where we know we are working with a single well known regnode type.

demerphq committed:
regen/regcomp.pl - create typedefs for all regnode types

Currently we hard code the struct used by the different regop types.
This makes it awkward to change the structure used by a specific regop
as the struct it uses might be used in many contexts, and each cases
of a regop using that structure must be reviewed to see if it needs
to be changed.

This patch adds a typedef for each regnode. The typedefs are named
'tregnode_OP', for instance 'tregnode_TRIE' is typedefed to 'struct
charclass' (at the time of this commit). This allows the code to do
things like 'sizeof(tregnode_TRIE)' and should the exact struct used
for TRIE regops change in the future then no code need be reviewed
or changed.

demerphq committed:
regen/regcomp.pl - Make regarglen available as PL_regarglen in regexec.c

In a follow up patch we will use this data from regexec.c which
currently cannot see the variable.

This changes a comment in regen/mk_invlists.pl which necessitated
rebuilding several files related to unicode. Only the hashes associated
with mk_invlists.pl were changed.

demerphq committed:
regcomp.c - rename NEXTOPER to REGNODE_AFTER and related logic

It is really easy to get confused about the difference between
NEXTOPER() and regnext() of a regnode. The two concepts are related,
similar, but importantly distinct. NEXTOPER() is also defined in such a
way that it is easy to abuse and misunderstand and encourages producing
code that is fragile to larger change, effectively "baking in"
assumptions to the code that are difficult to discover by searching.
Changing the type and storage requirements of a regnode may break things
in subtle and hard to debug ways.

An example of how NEXTOPER() is problematic is that this:
NEXTOPER(NEXTOPER(branch)) does not mean "find the second node after the
branch node", it means "jump forward by a regnode which happens to be
two regnodes large". In other words NEXTOPER is just a fancy way of
writing "node+1".

This patch replaces NEXTOPER() with three new macros:

    REGNODE_AFTER_dynamic(node)
    REGNODE_AFTER_opcode(node,op)
    REGNODE_AFTER_type(node,tregnode_OPNAME)

The first is the most generic case, it jumps forward by the size of the
node, and determines that size by consulting OP(node). The second is
where you have already extracted OP(node), and the third is where you
know the actual structure that you want to jump forward by. Every
regnode type has a corresponding type, which is known at compile time,
so using the third will produce the most efficient code. However in many
cases the code operates on one of several types, whose size may be the
same now, but may change in the future, in which case one of the other
forms is preferred. The run time logic in regexec.c should probably
only use the REGNODE_AFTER_type() interface.

Note that there is also a REGNODE_BEFORE() which replaces PREVOPER(),
which is used in a specific piece of legacy logic but should not be
used otherwise. It is not safe to go backwards from an arbitrary node,
we simply have no way to know how large the previous node is and thus
where it starts.

This patch includes some logic that validates assumptions during DEBUG
mode which should catch errors from resizing regnodes.

After this patch changing the size of an existing regnode should be
relatively safe and errors related to sizing should trigger assertion
fails.

This patch includes changes to perlreguts.pod to explain this stuff
better.

demerphq committed:
regcomp.c - replace repeated OP(n) with variable 'op'.

Declutter code.

demerphq committed:
regcomp.c - replace OP(n) macro with variable 'op' in S_dumpuntil()

This declutters the code and allows us to remove the casting as well.

As a byproduct the loop control logic is a bit simplified.

demerphq committed:
regen/mk_invlists.pl - under DEBUG=1 show some progress output

demerphq committed:
regex engine rename -> reg_off_by_arg

demerphq committed:
regex engine - Rename reg_off_by_arg to PL_reg_off_by_arg

This is in preparation for a future patch, so we can access
PL_reg_off_by_arg() from an inline function in regexec.c

demerphq committed:
regcomp.pl - use the regnode typedefs in EXTRA_SIZE calculations

demerphq committed:
regen/regcomp.pl - fix documentation (add missing PL_ prefix)

demerphq committed:
regen/regcomp.pl - add a way to dump the node/state table

For debugging and enhancements, etc.

demerphq committed:
regen/regcomp.pl - add PL_regargvaries

demerphq committed:
regex engine - Rename PL_regarglen to PL_regnode_arg_len

demerphq committed:
regex engine - Rename PL_regargvaries to PL_regnode_arg_len_varies

demerphq committed:
regex engine - Rename PL_regkind to PL_regnode_kind

demerphq committed:
regex engine - Rename PL_reg_off_by_arg to PL_regnode_off_by_arg

demerphq committed:
regex engine - Rename PL_reg_name to PL_regnode_name

demerphq committed:
globvar.sym - sort PL_reg*

demerphq committed:
regcomp.c - initial support for EXACTish nodes in regnode_after()

This is a first step, we add the support for EXACTish nodes
here, but we do not use it. In a following commit we will move
it to a new file. This patch is just to keep the move clean.

demerphq committed:
regex engine - rename REGNODE_AFTER_dynamic() REGNODE_AFTER()

Now that REGNODE_AFTER() can handle all cases it makes sense
to remove the dynamic() suffix.

demerphq committed:
regex engine - integrate regnode_after() support for EXACTish nodes

This adds REGNODE_AFTER_varies() which is used when the called *knows*
that the current regnode is variable length. We then use it to handle
EXACTish style nodes as determined by PL_regnode_arg_len_varies.

As part of this patch Perl_regnext() Perl_regnode_after() and
Perl_check_regnode_after() are moved to reginline.h, which is loaded via
regcomp.c only when we are compiling the regex engine.

jkeenan commented:
I see several branches with similar names.  How does yours compare to the other branches listed here:
```
$ git branch -a | grep regnode_typedef
  remotes/bram-perl/yves/regnode_typedefs
  remotes/origin/yves/regnode_typedefs
  remotes/pengvado/yves/regnode_typedefs
```

jkeenan commented on pod/perlreguts.pod:
```diff
@@ -216,39 +216,113 @@ types.
-it is important to keep them clear.
+There are two distinct concepts of "next regnode" in the regex engine,
+and it is important to keep them distinct in your thinking as they
+overlap conceptually in many places, but where they dont overlap the
```
missing an apostrophe; s/dont/don't/

jkeenan commented on pod/perlreguts.pod:
```diff
@@ -216,39 +216,113 @@ types.
 
 =back
 
+Most regnode types do not create a branch in the exection flow, and
```
typo: should be `exception`

demerphq replied:
execution actually. :-)

jkeenan commented on regcomp.h:
```diff
@@ -438,11 +438,68 @@ struct regnode_ssc {
+ * a function which is more powerful called Perl_regnode_after(), see
+ * regcomp.c.
+ *
+ * Note that regnodes can be thought of as nodes in binary tree like structure.
```
Recommend a hyphen between `tree` and `like`:  `tree-like`

khwilliamson commented on regcomp.h:
```diff
@@ -438,11 +438,68 @@ struct regnode_ssc {
+ * The first "child" of a node is its immediate successor in the program, the
+ * second "child" of a node, if it has one, will be stored as an index in the
+ * regnode struct, usually in a field marked next_off (but sometimes somewhere
+ * else).  To access the immediate successor of a reegnode you should use one
```
typo: regnode

khwilliamson commented on regcomp.h:
```diff
@@ -438,11 +438,68 @@ struct regnode_ssc {
+ * REGNODE_AFTER_opcode(NODE,OPCODE).
+ *
+ */
+#define	_REGNODE_AFTER_plus(p,extra)    ((p) + NODE_STEP_REGNODE + (extra))
```
We probably shouldn't be adding non-conforming symbol names.  `REGNODE_AFTER_PLUS_` should do

khwilliamson commented on regcomp.h:
```diff
@@ -438,11 +438,68 @@ struct regnode_ssc {
+ *
+ */
+#define	_REGNODE_AFTER_plus(p,extra)    ((p) + NODE_STEP_REGNODE + (extra))
+/* under DEBUGGING we check that all REGNODE_AFTER style macro did the
```
*macros

khwilliamson reviewed (approved):
Thank you for this.  

I only found trivial issues.

demerphq commented:
On Sat, 30 Jul 2022 at 21:29, James E Keenan ***@***.***>
wrote:

> I see several branches with similar names. How does yours compare to the
> other branches listed here:
>
> $ git branch -a | grep regnode_typedef
>   remotes/bram-perl/yves/regnode_typedefs
>   remotes/origin/yves/regnode_typedefs
>   remotes/pengvado/yves/regnode_typedefs
>
> Those are other peoples copies of the branch which they have pulled into
their repo, and then you have pulled from them into yours. Likely theirs is
old. The canonical branch is origin/yves/regnode_typesdef.

Yves
-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

demerphq committed:
regex engine - improved comments explaining REGNODE_AFTER()

This rewrites one comment to include more explanation of the difference
between Perl_regnext() and REGNODE_AFTER().

demerphq commented:
Hi, I integrated and rebased the requested changes. I also added a new comment with more and better docs. No code changes other than requested in the rebase. Please review the very last commit in the PR.

demerphq commented:
A follow up question from this PR for me is whether we should unify the regnode information data into a structure and replace all the PL_regnode_... arrays with something unified. We could even use packed integers and reduce the memory footprint involved, not sure if that would be a win or not.

demerphq commented:
@jkeenan and @khwilliamson need approvals from you guys again.
